server/aesdsocket.c

```diff
@@ -0,0 +1,379 @@
+    {
+        perror("bind");
+        syslog(LOG_ERR, "Error binding socket: %m");
+        close(sockfd);
+        return ERROR_SOCKET_BIND;
+    }
+
+    if (listen(sockfd, MAX_BACKLOG) == -1)    //setting the socket to the listening state using the listen function
+    {
+        perror("listen");
+        syslog(LOG_ERR, "Error listening on socket: %m");
+        close(sockfd);
+        return ERROR_LISTEN;
+    }
+
+    return sockfd;
+}
+
+
+/*----------------------------------------------------------------------------
+static void handle_client_connection(int client_fd) 
+ *@brief : Function to receive and send data from and to the client
+ *
+ *Parameters : client_fd
+ *
+ *Return : none
+ *
+ *----------------------------------------------------------------------------*/
+static void handle_client_connection(int client_fd) 
+{
+    char client_ip[MAX_IP_LEN];                   //to store information about the client's IP address. 
+    struct sockaddr_in client_addr;               //structure to hold client socket address information
+    socklen_t client_len = sizeof(client_addr);
+
+    getpeername(client_fd, (struct sockaddr *)&client_addr, &client_len);  //retrieve the IP address of the connected client
+    
+    //extracting the client's IP address from the client_addr structure and converting it from binary form to a string using inet_ntop.
+    inet_ntop(AF_INET, &(client_addr.sin_addr), client_ip, MAX_IP_LEN);
+
+    syslog(LOG_INFO, "Accepted connection from %s", client_ip);
+
+    char *buffer = (char *)malloc(sizeof(char) * BUFFER_SIZE);   //buffer used to read and write data to and from the client
+    if (buffer == NULL) 
+    {
+        syslog(LOG_ERR, "Could not allocate memory");
+        exit(ERROR_MEMORY_ALLOC);
+    } else 
+    {
+        syslog(LOG_ERR, "Successfully created buffer\n");
+    }
+
+    //opening a data file (DATA_FILE) in write-only mode with options to create the file if it doesn't exist and to append data to it.
+    int data_fd = open(DATA_FILE, O_WRONLY | O_CREAT | O_APPEND, FILE_PERMISSIONS);
+    if (data_fd == -1) 
+    {
+        perror("open");
+        syslog(LOG_ERR, "Error opening data file: %m");
+        free(buffer);
+        close(client_fd);
+        return;
+    }
+
+    ssize_t bytes_received;
+
+    //reading data from the client socket in chunks of up to BUFFER_SIZE bytes using the recv function.
+    while ((bytes_received = recv(client_fd, buffer, BUFFER_SIZE, 0)) > 0)  
+    {
+        write(data_fd, buffer, bytes_received);           //writing the received data to the data file using the write function.
+        if (memchr(buffer, '\n', bytes_received) != NULL) //checking if the received data contains a newline character 
+        {
+            break;
+        }
+    }
+
+    syslog(LOG_INFO, "Received data from %s: %s, %d", client_ip, buffer, (int)bytes_received);
```

Comment:
@vidhyapalaniappan it's not safe to use `%s` with `buffer` without ensuring null termination.  This is what the valgrind errors are pointing out.